File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int n = heights.size();
        int p1 = 0, p2 = n-1, max_water = 0;
        while(p1 < p2){
            int water = (p2 - p1) * min(heights[p1], heights[p2]);
            max_water = max(max_water, water);
            if(heights[p1] > heights[p2])
                p2--;
            else if(heights[p1] <= heights[p2])
                p1++;
        }
        return max_water;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> st = new Stack<>();

        int len = heights.length;
        int maxi = 0,pse=0;

        for(int i = 0; i < len; i++){
            if(!st.empty()){
                while(!st.empty() && heights[st.peek()] >= heights[i]){
                    int top = st.pop();
                    if(st.empty())
                        pse = -1;
                    else
                        pse = st.peek();
                    maxi = Integer.max(maxi, heights[top]*(i- pse -1));
                }
                st.push(i);
            }else{
                st.push(i);
            }
        }

        while(!st.empty()){
            int top = st.pop();
            if(st.empty())
                pse = -1;
            else
                pse = st.peek();
            maxi = Integer.max(maxi, heights[top]*(len - pse -1));
        }

        return maxi;
    }
}
